### PR TITLE
Reintroducing PostgreSQL Database Support in Mautic 7.x

### PR DESCRIPTION
**Pull request by wieslawgolec**
**This is a Draft PR** - Expecting significant review feedback, additional test cases, and possible refinements before marking Ready for Review.

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | ❌
| New feature/enhancement? (use the a.x branch) | ✔️
| Deprecations?                          | ❌
| BC breaks? (use the c.x branch)        | ❌
| Automated tests included?              | ✔️
| Related user documentation PR URL      | (TBD – will create after merge; suggests adding PostgreSQL to requirements/install docs)
| Related developer documentation PR URL | (TBD – will create after merge; notes on query compatibility layer)
| Issue(s) addressed                     |Reintroduces PostgreSQL support (see forum PoC https://forum.mautic.org/t/proof-of-concept-restoring-postgresql-support-for-mautic/36313 and roadmap proposal https://community.mautic.org/processes/roadmap/f/4/proposals/200)

## Description

This PR reintroduces **official PostgreSQL database support** in Mautic 7.x as a **minimal compatibility layer**, while preserving full compatibility and identical behavior with MySQL and MariaDB. No breaking changes for existing users.

### Key Changes & Fixes
- **Fully functional installer** - Works identically across PostgreSQL, MySQL, and MariaDB.
- All **migrations** adjusted to execute consistently on every supported database engine.
- Existing **unit & functional tests** updated and now passing reliably on PostgreSQL (as well as MySQL/MariaDB).
- Most **report & analytics queries** converted to standard SQL (removing MySQL-specific syntax where possible).
- **Column/alias quoting** added extensively in raw queries and QueryBuilder expressions - PostgreSQL lowercases unquoted identifiers automatically, while Mautic relies heavily on `camelCase` aliases.
- **Consistent string matching** across databases: MySQL/MariaDB `LIKE` is case-insensitive → uses `ILIKE` on PostgreSQL or `LOWER()`-wrapped comparisons as needed.
- Platform differences in standard **QueryBuilder** and **DQL** behavior resolved for uniform logic execution.
- **Report builder enhancement**: Automatically corrects/adds missing GROUP BY fields based on SELECT clauses (PostgreSQL enforces strict GROUP BY rules; this prevents errors without changing MySQL/MariaDB behavior).

### Testing & Compatibility Verified
Latest commits tested via GitHub Actions on:
- PHP 8.2-8.5
- PostgreSQL 16 & 18
- MariaDB 10.11 & 11.4
- MySQL 8.4 & 9.4

**Supported ranges**: PostgreSQL 16+, PHP 8.2+.

Core features (installer, contacts, segments, campaigns, emails, tracking, reports, etc.) verified end-to-end on PostgreSQL with no regressions on MySQL/MariaDB.

---

### 📋 Steps to test this PR:

1. Pull this PR (or use the source branch `7.x-postgresql` from https://github.com/wieslawgolec/mautic if submitting from fork).
2. Set up a fresh PostgreSQL 16+ database and configure credentials in `.env` (or `local/config.php`).
3. Run the installer: `php bin/console mautic:install` - complete a full fresh install.
4. Confirm all migrations apply cleanly: `php bin/console doctrine:migrations:migrate`.
5. Test key areas for PostgreSQL-specific issues:
   - Create contacts, manual/dynamic segments
   - Build/run campaigns with conditions/filters
   - Create and run multiple reports (especially those with aggregates, custom fields, GROUP BY, or LIKE filters)
   - Send test emails and verify tracking/pixel hits
   - Use search/filters in contact/lead lists that use LIKE/ILIKE
6. Repeat steps on a **MySQL** or **MariaDB** install to ensure zero regressions.
7. Run the full test suite: `./bin/phpunit` - verify it passes on both PostgreSQL and MySQL/MariaDB.
8. Specifically check reports: Create a report with SELECT columns not in GROUP BY (e.g., aggregated stats by month) - confirm it runs without "column must appear in GROUP BY" errors on PostgreSQL (and still works normally on MySQL/MariaDB).

**Note for testers**: Focus extra attention on reports, segments with complex filters, and any LIKE-based searches. Report any edge cases, performance differences, or remaining non-standard queries.

Feedback, additional test scenarios, and code review very welcome - this is Draft for a reason.

### CODE REVIEW

File: app/bundles/LeadBundle/Form/Type/FilterTrait.php [hunk 1]
```diff
@@ -317,10 +317,12 @@ public function buildFiltersForm($eventName, FormEvent $event, TranslatorInterfa
...
                             try {
-                                $qb = $this->connection->createQueryBuilder();
+                                $qb             = $this->connection->createQueryBuilder();
+                                $isPg           = $this->connection->getDatabasePlatform() instanceof \Doctrine\DBAL\Platforms\PostgreSQLPlatform;
+                                $whereCondition = $isPg ? 'CAST(l.id AS TEXT) ~* :regex' : 'l.id REGEX :regex';
                                 $qb->select('l.id')
                                     ->from(MAUTIC_TABLE_PREFIX.'leads', 'l')
-                                    ->where('l.id REGEXP :regex')
+                                    ->where($whereCondition)
                                     ->setParameter('regex', $this->prepareRegex($regex))
```
[the remark this answers is not in the document]
Comment (wieslawgolec) [reply app/bundles/LeadBundle/Form/Type/FilterTrait.php:336]: https://dev.mysql.com/doc/refman/8.0/en/regexp.html

**Comment by RCheesley** [comment]
Hi @wieslawgolec did you want folks to review this yet, or are you still working on it?

**Comment by wieslawgolec** [comment]
> Hi @wieslawgolec did you want folks to review this yet, or are you still working on it?

Hey @RCheesley,
This PR has actually been ready for review for quite a while now.
The long list of commits is mostly because I've been periodically merging the latest changes from upstream (master/7.x) and fixing whatever small incompatibilities popped up along the way. Nothing major, just keeping it in sync so it doesn't break when things change upstream.
The PostgreSQL support itself has been stable for some time now. 
All tests are green on MySQL, MariaDB, and PostgreSQL.
I'm leaving on a two-week trip tomorrow, so my replies might be a bit delayed during that time. Anyway I'm ready to address any feedback.

**Comment by wieslawgolec** [comment]
> ## [Codecov](https://app.codecov.io/gh/mautic/mautic/pull/15926?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic) Report
> 
> ❌ Patch coverage is `71.37255%` with `438 lines` in your changes missing coverage. Please review. ✅ Project coverage is 69.26%. Comparing base ([`8e323de`](https://app.codecov.io/gh/mautic/mautic/commit/8e323de1289d107106c27254a9f4bdc0ce78376c?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic)) to head ([`aa7fad1`](https://app.codecov.io/gh/mautic/mautic/commit/aa7fad1fa54d66be39efe7946c44550d691cd39e?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic)).
> [Files with missing lines](https://app.codecov.io/gh/mautic/mautic/pull/15926?dropdown=coverage&src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic) 	Patch % 	Lines
> [...llBundle/EventListener/DoctrineEventSubscriber.php](https://app.codecov.io/gh/mautic/mautic/pull/15926?src=pr&el=tree&filepath=app%2Fbundles%2FInstallBundle%2FEventListener%2FDoctrineEventSubscriber.php&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic#diff-YXBwL2J1bmRsZXMvSW5zdGFsbEJ1bmRsZS9FdmVudExpc3RlbmVyL0RvY3RyaW5lRXZlbnRTdWJzY3JpYmVyLnBocA==) 	6.45% 	[58 Missing ⚠️ ](https://app.codecov.io/gh/mautic/mautic/pull/15926?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic)
> [app/bundles/CoreBundle/Entity/CommonRepository.php](https://app.codecov.io/gh/mautic/mautic/pull/15926?src=pr&el=tree&filepath=app%2Fbundles%2FCoreBundle%2FEntity%2FCommonRepository.php&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic#diff-YXBwL2J1bmRsZXMvQ29yZUJ1bmRsZS9FbnRpdHkvQ29tbW9uUmVwb3NpdG9yeS5waHA=) 	43.01% 	[53 Missing ⚠️ ](https://app.codecov.io/gh/mautic/mautic/pull/15926?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic)
> [app/bundles/InstallBundle/Helper/SchemaHelper.php](https://app.codecov.io/gh/mautic/mautic/pull/15926?src=pr&el=tree&filepath=app%2Fbundles%2FInstallBundle%2FHelper%2FSchemaHelper.php&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic#diff-YXBwL2J1bmRsZXMvSW5zdGFsbEJ1bmRsZS9IZWxwZXIvU2NoZW1hSGVscGVyLnBocA==) 	44.06% 	[33 Missing ⚠️ ](https://app.codecov.io/gh/mautic/mautic/pull/15926?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic)
> [...s/CoreBundle/Doctrine/Helper/IndexSchemaHelper.php](https://app.codecov.io/gh/mautic/mautic/pull/15926?src=pr&el=tree&filepath=app%2Fbundles%2FCoreBundle%2FDoctrine%2FHelper%2FIndexSchemaHelper.php&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic#diff-YXBwL2J1bmRsZXMvQ29yZUJ1bmRsZS9Eb2N0cmluZS9IZWxwZXIvSW5kZXhTY2hlbWFIZWxwZXIucGhw) 	55.00% 	[27 Missing ⚠️ ](https://app.codecov.io/gh/mautic/mautic/pull/15926?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic)
> [.../bundles/LeadBundle/Entity/LeadFieldRepository.php](https://app.codecov.io/gh/mautic/mautic/pull/15926?src=pr&el=tree&filepath=app%2Fbundles%2FLeadBundle%2FEntity%2FLeadFieldRepository.php&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic#diff-YXBwL2J1bmRsZXMvTGVhZEJ1bmRsZS9FbnRpdHkvTGVhZEZpZWxkUmVwb3NpdG9yeS5waHA=) 	55.73% 	[27 Missing ⚠️ ](https://app.codecov.io/gh/mautic/mautic/pull/15926?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic)
> [app/bundles/LeadBundle/Entity/LeadRepository.php](https://app.codecov.io/gh/mautic/mautic/pull/15926?src=pr&el=tree&filepath=app%2Fbundles%2FLeadBundle%2FEntity%2FLeadRepository.php&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic#diff-YXBwL2J1bmRsZXMvTGVhZEJ1bmRsZS9FbnRpdHkvTGVhZFJlcG9zaXRvcnkucGhw) 	42.55% 	[27 Missing ⚠️ ](https://app.codecov.io/gh/mautic/mautic/pull/15926?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic)
> [...p/bundles/LeadBundle/Entity/ListLeadRepository.php](https://app.codecov.io/gh/mautic/mautic/pull/15926?src=pr&el=tree&filepath=app%2Fbundles%2FLeadBundle%2FEntity%2FListLeadRepository.php&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic#diff-YXBwL2J1bmRsZXMvTGVhZEJ1bmRsZS9FbnRpdHkvTGlzdExlYWRSZXBvc2l0b3J5LnBocA==) 	41.17% 	[20 Missing ⚠️ ](https://app.codecov.io/gh/mautic/mautic/pull/15926?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic)
> [...p/bundles/EmailBundle/Stats/EmailPeriodMetrics.php](https://app.codecov.io/gh/mautic/mautic/pull/15926?src=pr&el=tree&filepath=app%2Fbundles%2FEmailBundle%2FStats%2FEmailPeriodMetrics.php&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic#diff-YXBwL2J1bmRsZXMvRW1haWxCdW5kbGUvU3RhdHMvRW1haWxQZXJpb2RNZXRyaWNzLnBocA==) 	76.31% 	[18 Missing ⚠️ ](https://app.codecov.io/gh/mautic/mautic/pull/15926?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic)
> [...s/CampaignBundle/Entity/LeadEventLogRepository.php](https://app.codecov.io/gh/mautic/mautic/pull/15926?src=pr&el=tree&filepath=app%2Fbundles%2FCampaignBundle%2FEntity%2FLeadEventLogRepository.php&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic#diff-YXBwL2J1bmRsZXMvQ2FtcGFpZ25CdW5kbGUvRW50aXR5L0xlYWRFdmVudExvZ1JlcG9zaXRvcnkucGhw) 	69.81% 	[16 Missing ⚠️ ](https://app.codecov.io/gh/mautic/mautic/pull/15926?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic)
> [...p/bundles/CampaignBundle/Entity/LeadRepository.php](https://app.codecov.io/gh/mautic/mautic/pull/15926?src=pr&el=tree&filepath=app%2Fbundles%2FCampaignBundle%2FEntity%2FLeadRepository.php&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic#diff-YXBwL2J1bmRsZXMvQ2FtcGFpZ25CdW5kbGUvRW50aXR5L0xlYWRSZXBvc2l0b3J5LnBocA==) 	36.36% 	[14 Missing ⚠️ ](https://app.codecov.io/gh/mautic/mautic/pull/15926?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic)
> ... and [42 more](https://app.codecov.io/gh/mautic/mautic/pull/15926?src=pr&el=tree-more&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic) 		
> Additional details and impacted files
> 
> 		
> 		
> 		
> 		
> 		
> 		
> 		
> 		
> 		
> 		
> 		
> 		
> 
> 🚀 New features to boost your workflow:



> [Codecov](https://app.codecov.io/gh/mautic/mautic/pull/15926?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=mautic) Report

The coverage drop comes mostly from the new/changed database abstraction code (schema helpers, repositories, Doctrine subscriber, etc.) that was added for PostgreSQL compatibility.
These parts are already exercised by the existing test suite. All unit + functional tests now run and pass on PostgreSQL (as well as MySQL/MariaDB) in CI.
Anyway, I agree with the report, that some of the PostgreSQL-specific edge cases aren't explicitly covered by dedicated unit tests yet.
I'm not super keen on adding a bunch of PostgreSQL-only tests (it would duplicate a lot and make maintenance heavier).
Instead, I'd rather prefer to extend the existing tests to be platform-aware,  so they automatically adjust their assertions or data setup depending on whether they're running against MySQL, MariaDB or PostgreSQL.
That way we get better coverage for everyone without exploding the test count, which is already not a small number.

**Comment by RCheesley** [comment]
Thanks - the PR was still on draft, so it doesn't get onto people's radars. I just marked it ready for review and will tag it up as well. Thanks for all your work on this!

File: app/bundles/ApiBundle/Controller/FetchCommonApiController.php [hunk 2]
```diff
@@ -448,36 +448,49 @@ protected function getBatchEntities($parameters, &$errors, $prepareForSerializat
...
-        // Set errors
-        if ($idHelper->hasErrors()) {
-            foreach ($idHelper->getErrors() as $key => $error) {
-                $this->setBatchError($key, $error, Response::HTTP_BAD_REQUEST, $errors);
-            }
+        // Set errors for invalid or not-found IDs
+        $foundIds = [];
+        foreach ($entities as $entity) {
+            $foundIds[$entity->getId()] = true;
         }
```
[the remark this answers is not in the document]
Comment (wieslawgolec) [reply app/bundles/ApiBundle/Controller/FetchCommonApiController.php]: This dead code with $foundIds was removed and replaced with TODO comment for proper implementation in the future. It looks like an obvious typo at first glance, but this one is actually intentional (but unfinished).
I deliberately wrote it this way to keep the exact same behavior across both MySQL (mautic traditional DB) and PostgreSQL. MySQL silently ignores/converts certain invalid inputs in these contexts, while PostgreSQL is stricter and throws an error. This tweak makes Postgres mimic MySQL's more lenient handling so we don't break existing setups or installer flows when people run on Postgres.
It's was suppose to be compatibility shim.

File: app/bundles/ApiBundle/Controller/FetchCommonApiController.php [hunk 3]
```diff
@@ -448,36 +448,49 @@ protected function getBatchEntities($parameters, &$errors, $prepareForSerializat
...
+        // Extract all requested IDs (including invalid ones for error reporting)
+        $requestedIds = $idHelper->getIds(); // associative: original_key => id_value
+
+        // Filter only valid positive integer IDs for the query
+        $validIds = [];
+        foreach ($requestedIds as $key => $id) {
+            if (is_numeric($id) && (int) $id > 0 && (string) (int) $id === (string) $id) {
+                $validIds[$key] = (int) $id;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (wieslawgolec) [reply app/bundles/ApiBundle/Controller/FetchCommonApiController.php:462]: This isn't a mistake. It's on purpose to preserve consistent behavior between MySQL and PostgreSQL environments.
MySQL tends to silently swallow or adjust invalid input in these cases, but PostgreSQL is much stricter and will error out. The code is written this way specifically so Postgres behaves like MySQL here and doesn't introduce new failures for users.
It's a small compatibility bridge we need until we can clean up the DB layer more broadly.

**Comment by wieslawgolec** [comment]
> Thanks - the PR was still on draft, so it doesn't get onto people's radars. I just marked it ready for review and will tag it up as well. Thanks for all your work on this!

@RCheesley
Thanks for checking in!
It's still a draft because I expect some refactoring and polishing once I get feedback from the maintainers as my coding style isn't fully aligned with Mautic's yet.
The core logic is already complete and all tests are passing on MySQL, MariaDB, and PostgreSQL.
That said, I'd really appreciate early reviews as it is right now. Any suggestions on structure, moving logic into helper classes/traits, simplifications, or following Mautic's preferred patterns would be very welcome before I do the final cleanup and mark it as "Ready for Review".
Feel free to push commits directly to this branch if you see something that can be improved - optimizations are definitely still possible! 👍

**Comment by RCheesley** [comment]
Some conflicts to resolve here @wieslawgolec - sorry about that! I'll see if we can get some 👀 on this so we can merge it in sooner rather than later!

**Comment by RCheesley** [comment]
@promptless-for-oss can you make a docs PR for this one please. We need one PR for user docs and one PR for the developer docs.

**Comment by escopecz** [comment]
I was quite excited about the [proposal](https://community.mautic.org/processes/roadmap/f/4/proposals/200) as it was suggesting to bend Postgres with polyfills to support Mautic rather than bending Mautic to support Postgres. However, in this PR I can see major Mautic refactoring. Lots of ifs where one query is used for MySql and MariaDB and completely different query is used for Postgres. When you look at the size of this PR it's clear how massive change it is. It'll take weeks just to review this. I'm pretty sure there are tens of bugs hidden in this PR. So we'll need good user base that require Mautic running on Postgres that will help us review and test this PR. Because if it is still just 1% then such a massive change that makes the code even harder to maintain isn't worth it. This is why we've decided to drop support for Posgres in the first place.

**Review by shinde-rahul — CHANGES_REQUESTED** [review]
I’m aligned with @escopecz  here.

The proposal is exciting because it aims to add a compatibility layer while keeping Mautic mostly unchanged. This PR feels like it’s going in the opposite direction; we’re introducing a lot of refactoring and DB-specific branching across the codebase.

That does raise some concerns,

* DB-specific logic is starting to leak into the application layer
* Similar queries are being implemented differently per database
* The size of the PR makes it harder to review thoroughly and increases the risk of regressions


For example, patterns like this start to add up quickly:

```php
if ($isPostgres) {
    $query = 'SELECT ... PG specific ...';
} else {
    $query = 'SELECT ... MySQL specific ...';
}
```

Or duplicating logic per DB:

```php
public function getContactsMysql() { ... }
public function getContactsPostgres() { ... }
```

Or inline SQL differences:

```php
$sql = 'SELECT DATE_FORMAT(created_at, ...)';
if ($isPostgres) {
    $sql = 'SELECT TO_CHAR(created_at, ...)';
}
```

Individually, these may seem manageable, but over time, they can make the code harder to reason about, maintain, and extend, especially as more edge cases emerge.

It might scale better to introduce a single abstraction point for these differences, for example:

```php
$query = $this->dialect->buildQuery('some_operation', $params);
```

Or even handle more of this at the DB/platform layer where possible, so the application code remains consistent.

Some targeted, well-contained changes are absolutely reasonable. However, at this scale, it starts to feel like we’re shaping Mautic around Postgres, rather than adapting Postgres in a way that fits Mautic’s existing architecture.

It may be worth stepping back and realigning with the original proposal:

* Keep core logic DB-agnostic
* Introduce a clear compatibility/abstraction layer
* Avoid scattering conditionals across the codebase

The goal is solid and definitely worth pursuing; this just needs a tighter architectural direction to keep it maintainable for the community over time.

**Comment by dongilbert** [comment]
I've considered a better database abstraction layer for Mautic for some time. Inserting another abstraction layer for all data-driven interactions is very appealing. Such a mechanism would allow us to use different backend data sources  depending on what's being requested. Have some documents in MongoDB? Go get them through the data layer Mongo datasource. Need to operate on your campaigns from Salesforce? Get them through the data layer's Salesforce datasource. This would prevent us from having to manage the seemingly neverending data sync issues that come hand-in-hand with our current integrations. 

There's plenty of architectural benefits of this approach. By managing the data layer in this way, we could intercept and act on any request for data, modifying or otherwise combining various data sources into a normalized structure. We could take a request like `$this->datalayer->contacts->all()` and pull in local db contacts, remote contacts from an API response, and query the filesystem of an old ERP. We could more easily implement an [event sourcing](https://eventsauce.io/) approach which comes with all sorts of benefits on its own (data rollbacks, replays, insight into what happened for audit purposes). 

That said, such a system could be rather impractical both to implement and in real world usage without major changes that go beyond the scope of what even this PR is trying to accomplish. I'm not opposed to supporting pgsql, but I would like to see it implemented in a particular way if we really want it. I'm open to discussion.

**Comment by wieslawgolec** [comment]
Thanks for the detailed reviews, @escopecz @shinde-rahul @dongilbert. You're right to call out the scattered DB-specific logic and the growth in scope.

The original plan was a small, contained change like the POC: mostly platform tweaks, quoting fixes, and minimal adjustments to keep everything working on both MySQL/MariaDB and PostgreSQL without breaking existing installs. Doctrine 2.x limits cleaner abstraction options, so some conditionals and duplicated query parts crept in as we hit real edge cases in tests and production data. Things like LIKE vs ILIKE (with LOWER), REGEXP differences, date formatting (DATE_FORMAT vs TO_CHAR), GROUP BY strictness, and upsert behavior all needed handling. The PR ended up bigger than intended because covering those reliably took more work.

I won't promise a full cleanup in the next few days, but I will push the specific refactors you suggested: reduce the inline if ($isPostgres) branches where possible, centralize more of the dialect-specific building (following @shinde-rahul's idea of a clearer abstraction point), and clean up the duplicated methods.

All scattered platform checks and all shared code handling the differences between the PostgreSQL and MySQL dialects will be implemented in the DatabasePlatform class (single place to review), which looks like the best place for this purpose. This will eliminate the need for scattered If(platform) checks throughout the core code.

These reviews were actually expected and were exactly the reason this PR was being a draft. To get the working code reviewed early and gather feedback on the approach. The code itself works, but the current implementation is not yet at the production standard. That's why I wanted reviews before finalizing the architecture.

On maintenance: as mentioned before, I will personally handle PostgreSQL-related fixes going forward. I'm already running this fork in production (sending ~10 million emails/month), so any conflict, bug, or incompatibility that surfaces gets fixed on my side. No extra burden on the core team.

So the current plan is to:
- Short term (coming weeks): address the review comments, reduce scattering, improve query handling for the mentioned differences.
- After that: keep the PR focused on compatibility without major refactoring.
- Ongoing: I maintain Postgres fixes; core stays MySQL-first.

**Comment by wieslawgolec** [comment]
@escopecz Update (following @shinde-rahul's idea of a clearer abstraction point):

Consolidated platform-specific logic into the central **Mautic\CoreBundle\Doctrine\DatabasePlatform** class.
Remaining **DatabasePlatform::isPostgreSQL()** checks and why they stay:
1. **app/bundles/InstallBundle/EventListener/DoctrineEventSubscriber.php** → **postgreSqlMySqlCompact()** method - runs only on PostgreSQL during schema generation to create MySQL compatibility functions/operators (e.g. **date_format**, **convert_tz**, **ifnull**, boolean/integer operators). This is intentional and required for fresh installs.
2. **app/bundles/InstallBundle/Helper/SchemaHelper.php** → **installSchema()**, **backupExistingSchema()**, **dropExistingSchema()** - handles PostgreSQL-specific differences during installation like dropping sequences, backing up/renaming sequences, skipping **foreign_key_checks** (which does not exist on PostgreSQL) etc. These paths are required for the installer to work correctly on both platforms.
3. **app/bundles/CoreBundle/Test/MauticMysqlTestCase.php** - several methods use platform checks for transaction/rollback behavior, sequence handling, psql vs mysql commands, and dump/reset SQL generation etc. These are test infrastructure and deliberately platform-aware.
4. Tests:
* app/bundles/LeadBundle/Tests/EventListener/CampaignSubscriberFunctionalTest.php (testRegexOperatorOnDateFieldCondition) - keeps $useCleanupRollback disabled only for MySQL (to avoid "There is no active transaction" exception occur on MySQL - original test creator intention).
* app/bundles/CoreBundle/Tests/Functional/EventListener/MigrationCommandSubscriberTest.php - generated columns assertions on PostgreSQL (they are not immutable in the same way, so the test is marked as partly skipped).
5. 7 legacy migration files - old migrations (not executed on new installs) but must still run successfully when importing/upgrading from very old databases. They contain checks for:
* **foreign_key_checks** (MySQL-only)
* table field types
* charset/collation handling (PostgreSQL uses different syntax/behavior here)

Everywhere else in the codebase, scattered inline **isPostgreSQL()** checks have been removed and replaced with calls to the centralized **DatabasePlatform** helpers.

The remaining checks are isolated to:
* installation schema preparation
* legacy migration compatibility
* test-specific edge cases

**Comment by adiati98** [comment]
Hey @wieslawgolec, 

Promptless is suggesting changes for the user docs based on this PR.

Can you please review the content in [this docs PR](https://github.com/mautic/user-documentation/pull/697/changes) if it's correct and inline with the changes that have been made here?

TIA! ✨

**Comment by wieslawgolec** [comment]
> Hey @wieslawgolec,
> 
> Promptless is suggesting changes for the user docs based on this PR.
> 
> Can you please review the content in [this docs PR](https://github.com/mautic/user-documentation/pull/697/changes) if it's correct and inline with the changes that have been made here?
> 
> TIA! ✨

@adiati98 Done. Left you a comment there.

**Comment by escopecz** [comment]
Even after refactoring this is a massive change that will make the maintenance way more painful once we merge this. We are having troubles even supporting both MySql and MariaDB as there are more and more differences between the 2 over time.

This seem to me like a high cost low reward type of change. I like PosgreSQL. We started the Public Mautic Marketplace using Supabase. I'd like to use Supabase as Mautic DB as well. But it's still a "nice to have" that won't actually bring much value. Especially to the end user.

Can you please write what value this massive change will bring to Mautic?

**Comment by wieslawgolec** [comment]
> Even after refactoring this is a massive change that will make the maintenance way more painful once we merge this. We are having troubles even supporting both MySql and MariaDB as there are more and more differences between the 2 over time.
> 
> This seem to me like a high cost low reward type of change. I like PosgreSQL. We started the Public Mautic Marketplace using Supabase. I'd like to use Supabase as Mautic DB as well. But it's still a "nice to have" that won't actually bring much value. Especially to the end user.
> 
> Can you please write what value this massive change will bring to Mautic?

Hi @escopecz,
I get the concern. Maintaining DB differences is already tricky with MySQL/MariaDB, and this adds more surface area in the short term.

Pros:
- Opens Mautic to Postgres users (including Supabase, which you mentioned).
- Better standards compliance and stricter query behavior can catch bugs early.
- Community/enterprise interest exists (per forum).

Cons:
- Extra maintenance overhead now, especially with the old Doctrine version forcing heavy DatabasePlatform workarounds.
- Risk of subtle regressions if we miss edge cases.

That said, it's not urgent for production. I can keep it as an experimental/off-main-branch option for now. Most of the platform-specific code will vanish once we move to Doctrine 4.x anyway.
Can park it until the Doctrine upgrade if the cost/benefit doesn't feel right yet.

**Comment by escopecz** [comment]
@wieslawgolec all valid points! Let's focus on Doctrine upgrade then. Good to know there is a simpler way forward.

File: app/migrations/Version20260120130000.php [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+declare(strict_types=1);
+
+namespace Mautic\Migrations;
+
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+use Mautic\CoreBundle\Doctrine\DatabasePlatform;
+
+final class Version20260120130000 extends AbstractMigration
```
Comment (andersonjeccel) [inline app/migrations/Version20260120130000.php]: new migrations usually extend `AbstractMauticMigration` to follow the Mautic migration standard. worth switching this from Doctrine base `AbstractMigration` unless there is a specific reason this one cannot use the Mautic base.

Comment (wieslawgolec) [reply app/migrations/Version20260120130000.php]: switched to `AbstractMauticMigration`. Thanks

**Comment by thib-d** [comment]
We've been running this branch (`7.x-postgresql`) against PostgreSQL 16 and hit a reproducible failure on every `doctrine:migrations:migrate` run:

```
SQLSTATE[42P16]: multiple primary keys for table "migrations" are not allowed
```

**Root cause**: `app/config/config.php` applies `PDO::ATTR_STRINGIFY_FETCHES => true` to the connection options unconditionally (it's there as a MySQL/PHP 8.1 workaround). On `pdo_pgsql`, the option also stringifies the booleans returned by DBAL's schema-introspection queries against `pg_catalog` (e.g. `pg_index.indisprimary`/`indisunique`), which breaks DBAL's boolean handling — `introspectTable()` ends up reporting **zero indexes** for every table. doctrine-migrations then thinks the `migrations` table has no primary key and emits `ALTER TABLE migrations ADD PRIMARY KEY (version)` on every run → 42P16.

**Fix that works for us** — make the option conditional on the driver:

```php
'options' => 'pdo_pgsql' === $localConfigParameterBag->get('db_driver')
    ? []
    : [PDO::ATTR_STRINGIFY_FETCHES => true],
```

With that change, introspection is correct and `mautic:install` + all migrations run cleanly on PostgreSQL; we have the branch passing our end-to-end suite (install → login → dashboard) with it.

**Comment by wieslawgolec** [comment]
> ```
> ATTR_STRINGIFY_FETCHES
> ```

Hi @thib-d . Thanks for the detailed report and the suggested fix.

I've already removed the unconditional `PDO::ATTR_STRINGIFY_FETCHES` (no longer needed with PHP 8.2+ / MySQL 8.4+). 

Appreciate the help.

File: .github/workflows/tests.yml [hunk 5]
```diff
@@ -152,6 +167,173 @@
...
+        run: |
+          # 1. Extract major version number (e.g., '19beta1-alpine' -> '19')
+          VERSION=$(echo "${{ matrix.postgres-versions }}" | sed -E 's/([0-9.]+).*/\1/' | cut -d. -f1)
+
+          sudo apt-get update
+          sudo apt-get install -y curl ca-certificates gnupg lsb-release
+
+          # 2. Securely pull and dearmor the official PGDG signing key
+          sudo mkdir -p /usr/share/keyrings
+          curl -fSL https://www.postgresql.org/media/keys/ACCC4CF8.asc | gpg --dearmor | sudo tee /usr/share/keyrings/postgresql-archive-keyring.gpg > /dev/null
```
[the remark this answers is not in the document]
Comment (wieslawgolec) [reply .github/workflows/tests.yml:245]: false positive from AI (already https)
